File: backend/Api/Master/Master.Api/Services/MenuService.cs
using Master.Entity.Enums;
using Master.Entity.Models;
using Master.Entity.TableEntities;
using Master.Repository;
using Microsoft.EntityFrameworkCore;
using Shared.Kernel.Apps;
using Shared.Kernel.Tenancy;

namespace Master.Api.Services;

/// <summary>
/// Builds the navigation tree for whoever is asking.
///
/// The rows live in one self-referencing table, so this reads them flat in a single
/// query and assembles the three levels here rather than joining the same table
/// three times. There are on the order of a hundred of them; the shaping is cheaper
/// than the round trips.
///
/// A screen is offered only when the caller holds at least one of its permissions,
/// an empty section is dropped, and a module left with nothing to show is dropped in
/// turn — so the rail never draws an entry that leads nowhere. Rows with no route
/// are inactive in the seed and filtered here, which is how screens that are designed
/// but not yet built stay out of a user's way while remaining in the table.
///
/// A screen that belongs to another trade than the branch's is dropped too
/// (<see cref="TradeScope"/>, TK-30).
/// </summary>
public sealed class MenuService
{
    private readonly AdminDbContext _db;
    private readonly ITenantContext _tenant;

    public MenuService(AdminDbContext db, ITenantContext tenant)
    {
        _db = db;
        _tenant = tenant;
    }

    /// <summary>
    /// The caller's menu in one app (H0.3, TK-44): only rows whose
    /// <c>Menu.Apps</c> include it. A token that predates apps is RetailErp's.
    /// </summary>
    public async Task<IReadOnlyList<MenuView>> GetUserMenuAsync(CancellationToken ct, App app = App.RetailErp)
    {
        var permissions = _tenant.Permissions is { Count: > 0 }
            ? _tenant.Permissions
            : new HashSet<string>();

        // The branch's trade narrows the menu (D-10, TK-30): a screen that
        // belongs to another trade is not offered. General when unknown, which
        // shows everything — the same default the branch itself has.
        Guid? orgId = _tenant.OrgId;
        Vertical trade = orgId is null
            ? Vertical.General
            : await _db.Organizations.AsNoTracking()
                .Where(o => o.OrgId == orgId.Value)
                .Select(o => (Vertical?)o.Vertical)
                .FirstOrDefaultAsync(ct) ?? Vertical.General;

        var rows = await _db.Menus
            .AsNoTracking()
            .Where(m => m.IsActive && m.Apps.HasFlag(app))
            .OrderBy(m => m.DisplayOrder)
            .Select(m => new Row
            {
                MenuId = m.MenuId,
                ParentId = m.ParentId,
                Type = m.Type,
                Code = m.Code,
                Name = m.Name,
                Icon = m.Icon,
                Module = m.Module,
                RoutePath = m.RoutePath,
                IsSearchable = m.IsSearchable,
                CanCreate = m.CanCreate,
                SingularName = m.SingularName,
                AllowedActions = m.Permissions
                    .Where(p => permissions.Contains(p.PermissionCode))
                    .Select(p => p.Action)
                    .ToList()
            })
            .ToListAsync(ct);

        var childrenOf = rows
            .Where(r => r.ParentId is not null)
            .GroupBy(r => r.ParentId!.Value)
            .ToDictionary(g => g.Key, g => g.ToList());

        var menus = new List<MenuView>();

        foreach (var rail in rows.Where(r => r.Type == MenuType.Rail))
        {
            var groups = new List<MenuGroupView>();

            foreach (var group in Children(childrenOf, rail.MenuId).Where(r => r.Type == MenuType.Group))
            {
                var items = Children(childrenOf, group.MenuId)
                    .Where(r => r.Type == MenuType.Item && r.AllowedActions.Count > 0
                        && TradeScope.ShowsMenu(r.Code, trade))
                    .Select(r => new SubMenuView
                    {
                        SubMenuId = r.MenuId,
                        Code = r.Code,
                        Name = r.Name ?? r.Code,
                        RoutePath = r.RoutePath,
                        Icon = r.Icon,
                        Module = r.Module ?? string.Empty,
                        CanCreate = r.CanCreate,
                        SingularName = r.SingularName,
                        HasAccess = true,
                        AllowedActions = r.AllowedActions,
                    })
                    .ToList();

                if (items.Count == 0)
                {
                    continue;
                }

                groups.Add(new MenuGroupView
                {
                    MenuGroupId = group.MenuId,
                    Code = group.Code,
                    Name = group.Name,
                    Icon = group.Icon,
                    SubMenus = items,
                });
            }

            // A module with a route of its own — Home — has no sections to be
            // emptied, so it survives on its own permission rather than a child's.
            var reachable = groups.Count > 0
                || (rail.RoutePath is not null
                    && rail.Module is not null
                    && permissions.Contains($"{rail.Module}.view"));

            if (!reachable)
            {
                continue;
            }

            menus.Add(new MenuView
            {
                MenuId = rail.MenuId,
                Code = rail.Code,
                Name = rail.Name ?? rail.Code,
                Icon = rail.Icon,
                Module = rail.Module ?? string.Empty,
                RoutePath = rail.RoutePath,
                IsSearchable = rail.IsSearchable,
                Groups = groups,
            });
        }

        return menus;
    }

    private static List<Row> Children(IReadOnlyDictionary<int, List<Row>> index, int parentId) =>
        index.TryGetValue(parentId, out var kids) ? kids : [];

    /// <summary>One row as it comes back from the database, before it is shaped.</summary>
    private sealed class Row
    {
        public int MenuId { get; init; }
        public int? ParentId { get; init; }
        public MenuType Type { get; init; }
        public string Code { get; init; } = null!;
        public string? Name { get; init; }
        public string? Icon { get; init; }
        public string? Module { get; init; }
        public string? RoutePath { get; init; }
        public bool IsSearchable { get; init; }
        public bool CanCreate { get; init; }
        public string? SingularName { get; init; }
        public List<string> AllowedActions { get; init; } = [];
    }
}

File: frontend/libs/app-shell/src/lib/menu.service.ts
import { APP_ID } from '@bill-book/auth';
import { HttpClient } from '@angular/common/http';
import { Injectable, computed, inject, signal } from '@angular/core';
import { firstValueFrom } from 'rxjs';
import { MenuGroupView, MenuView, SubMenuView } from './menu.models';
import {
  SHELL_SCREENS,
  ShellRailItem,
  ShellScreen,
  ShellScreenGroup,
} from './shell-screens';

/**
 * The navigation tree, loaded once per session from `GET /api/menu`.
 *
 * The server decides what this user may see — which modules appear in the rail,
 * which screens hang under them, and which actions they hold on each. The client
 * draws what it is given; it does not re-filter, and it does not keep its own idea
 * of what screens exist.
 *
 * The one exception is the fallback. If the call fails — the endpoint is down, the
 * browser is offline, the response is unusable — the shell falls back to the static
 * registry in `shell-screens.ts` so the person can still navigate. That registry is
 * a safety net, not a second source of truth: when the call succeeds it is not
 * consulted at all.
 *
 * None of this is a security boundary. The rail is drawn from what the server sent,
 * the router guards on the same claims, and the server checks them again on every
 * request. That last check is the one that counts.
 */
@Injectable({ providedIn: 'root' })
export class MenuService {
  private readonly http = inject(HttpClient);
  private readonly app = inject(APP_ID);

  private readonly menus = signal<readonly MenuView[]>([]);

  /** True once a load has finished, whether it succeeded or fell back. */
  readonly loaded = signal(false);

  /** True when the tree on screen is the fallback rather than the server's. */
  readonly usingFallback = signal(false);

  readonly tree = computed(() => this.menus());

  /**
   * The rail, derived from the tree. A module with no route of its own lands on
   * the first screen in its panel — the panel is not built yet, so the rail has
   * to take the person somewhere.
   *
   * Empty until the tree loads; the components that draw a rail fall back to
   * `FALLBACK_RAIL` while it is.
   */
  readonly rail = computed<ShellRailItem[]>(() =>
    this.menus()
      .map((menu): ShellRailItem | null => {
        const firstScreen = menu.groups
          .flatMap((group) => group.subMenus)
          .find((sub) => !!sub.routePath)?.routePath;
        // A rail entry is where a module lands, so it drops any query the screen
        // carried: landing pre-filtered on one document type is not what clicking
        // the module means.
        const path = (menu.routePath ?? firstScreen)?.split('?')[0];
        return path
          ? { path, label: menu.name, icon: menu.icon ?? '', module: menu.module }
          : null;
      })
      .filter((item): item is ShellRailItem => item !== null),
  );

  /**
   * Every screen the person can reach, flattened — what Search and Favourites
   * read. Rows with no route are already filtered out by the server, but a null
   * slips through no worse than a broken link would, so guard here too.
   */
  readonly screens = computed<ShellScreen[]>(() =>
    this.menus().flatMap((menu) => {
      const direct: ShellScreen[] = menu.routePath
        ? [{ label: menu.name, path: menu.routePath, group: menu.name, module: menu.module }]
        : [];
      const nested = menu.groups.flatMap((group) =>
        group.subMenus
          .filter((sub) => !!sub.routePath)
          .map((sub) => ({
            label: sub.name,
            path: sub.routePath as string,
            group: menu.name,
            module: sub.module,
          })),
      );
      return [...direct, ...nested];
    }),
  );

  /** The same screens grouped for the panels that render headings. */
  readonly screenGroups = computed<ShellScreenGroup[]>(() =>
    this.menus()
      .map((menu) => ({
        name: menu.name,
        items: this.screens().filter((s) => s.group === menu.name),
      }))
      .filter((g) => g.items.length > 0),
  );

  /**
   * Load the tree. Safe to call more than once; the second call refreshes.
   * Never rejects — a failure lands on the fallback and says so.
   */
  async load(): Promise<void> {
    try {
      // The app is named for the logs and any cache between here and Master;
      // the token decides which app's rows come back (TK-44).
      const menus = await firstValueFrom(
        this.http.get<MenuView[]>('/api/menu', { params: { app: this.app } }),
      );
      if (!Array.isArray(menus) || menus.length === 0) {
        this.fallback();
        return;
      }
      this.menus.set(menus);
      this.usingFallback.set(false);
    } catch {
      this.fallback();
    } finally {
      this.loaded.set(true);
    }
  }

  /** The groups under one module, for its submenu panel. */
  groupsFor(menuCode: string): readonly MenuGroupView[] {
    return this.menus().find((m) => m.code === menuCode)?.groups ?? [];
  }

  /**
   * Which module a URL belongs to — what the secondary menu shows.
   *
   * Matched by longest path prefix rather than by first segment, because a module's
   * screens do not all live under one: reconciliation sits in Banking's panel while
   * its route is `/accounting/reconciliation`. The longest match wins so a screen is
   * attributed to the module that actually owns it.
   */
  menuForPath(url: string): MenuView | null {
    const clean = url.split('?')[0];
    let best: MenuView | null = null;
    let bestLength = 0;

    for (const menu of this.menus()) {
      const paths = [
        menu.routePath,
        ...menu.groups.flatMap((g) => g.subMenus.map((s) => s.routePath)),
      ].filter((p): p is string => !!p);

      for (const path of paths) {
        const candidate = path.split('?')[0];
        const matches = clean === candidate || clean.startsWith(`${candidate}/`);
        if (matches && candidate.length > bestLength) {
          best = menu;
          bestLength = candidate.length;
        }
      }
    }

    return best;
  }

  /**
   * The row that owns a route, when the tree knows about it.
   *
   * Some rows carry a query — `/sales/transactions?type=Invoice` is the invoice
   * register — so an exact match is tried first and the bare path only after.
   * Reversing that order would hand every typed register back as the mixed one.
   */
  subMenuFor(path: string): SubMenuView | null {
    const rows = this.menus().flatMap((menu) => menu.groups.flatMap((group) => group.subMenus));
    const exact = rows.find((s) => s.routePath === path);
    if (exact) return exact;

    // Only a row with no query of its own may match on the bare path. Without that
    // guard `/sales/transactions` would answer with whichever typed register came
    // first, since they all share that path.
    const clean = path.split('?')[0];
    return rows.find((s) => !!s.routePath && !s.routePath.includes('?') && s.routePath === clean) ?? null;
  }

  /**
   * Whether this user may take an action on a screen — `allows('/sales/invoices',
   * 'create')`. A screen the tree has never heard of answers false: the shell
   * should not invent a permission it was not given.
   */
  allows(path: string, action: string): boolean {
    return this.subMenuFor(path)?.allowedActions.includes(action) ?? false;
  }

  /**
   * Rebuild the tree from the static registry, shaped like the server's response
   * so nothing downstream has to know the difference. Every action is granted:
   * the fallback exists so a person is not stranded, and the server refuses
   * anything they should not do regardless of what the buttons offer.
   */
  private fallback(): void {
    const byGroup = new Map<string, ShellScreen[]>();
    for (const screen of SHELL_SCREENS) {
      const list = byGroup.get(screen.group) ?? [];
      list.push(screen);
      byGroup.set(screen.group, list);
    }

    let menuId = 0;
    let subId = 0;
    const menus: MenuView[] = [...byGroup.entries()].map(([name, screens]) => {
      menuId += 1;
      return {
        menuId,
        code: name.toLowerCase().replace(/[^a-z0-9]+/g, '-'),
        name,
        icon: null,
        module: screens[0]?.module ?? 'dashboard',
        routePath: screens.length === 1 ? screens[0].path : null,
        isSearchable: false,
        groups:
          screens.length === 1
            ? []
            : [
                {
                  menuGroupId: menuId,
                  code: `fallback-${menuId}`,
                  name: null,
                  icon: null,
                  subMenus: screens.map((screen) => {
                    subId += 1;
                    return {
                      subMenuId: subId,
                      code: `fallback-${subId}`,
                      name: screen.label,
                      routePath: screen.path,
                      icon: null,
                      module: screen.module ?? 'dashboard',
                      canCreate: false,
                      singularName: null,
                      hasAccess: true,
                      allowedActions: ['view', 'create', 'edit', 'delete', 'print', 'export'],
                    };
                  }),
                },
              ],
      };
    });

    this.menus.set(menus);
    this.usingFallback.set(true);
  }
}
